x-pack/test/functional_execution_context/test_utils.ts

```diff
@@ -41,12 +50,89 @@ export async function assertLogContains({
 }): Promise<void> {
   // logs are written to disk asynchronously. I sacrificed performance to reduce flakiness.
   await retry.waitFor(description, async () => {
-    const logsStr = await Fs.readFile(logFilePath, 'utf-8');
-    const normalizedRecords = logsStr
-      .split(endOfLine)
-      .filter(Boolean)
-      .map((s) => JSON.parse(s));
-
-    return normalizedRecords.some(predicate);
+    if (!logstream$) {
+      logstream$ = getLogstream$();
+    }
+    try {
+      await firstValueFrom(logstream$.pipe(filter(predicate), timeout(5_000)));
```

Comment:
Instead of reading the file over and over for every assertion, loop through the parsed log entries and await for the desired entry to occur if necessary 😇